sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -613,6 +613,15 @@ object LimitPushDown extends Rule[LogicalPlan] {
         case _ => join
       }
       LocalLimit(exp, newJoin)
+
+    case LocalLimit(limitExpr @ IntegerLiteral(limitVal),
+        window @ Window(Seq(Alias(WindowExpression(_: RankLike | _: RowNumber,
+            WindowSpecDefinition(Nil, orderSpec,
+                SpecifiedWindowFrame(RowFrame, UnboundedPreceding, CurrentRow))), _)), _, _, child))
+        if child.maxRows.forall( _ > limitVal) =>
+      LocalLimit(
+        limitExpr = limitExpr,
+        child = window.copy(child = Limit(limitExpr, Sort(orderSpec, true, child))))
```

Comment:
Wondering why do we need an extra `Sort` here? Shouldn't physical plan rule `EnsureRequirements` add the sort between window and limit?

Reply:
Yes, `Sort` is needed because we need global sort.

Reply:
If no partitionSpec specified, I think the planner inserts an exchange for making a single partition. So, we don't need a global sort here?

Reply:
The current logic is sort first and then limit:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Window [row_number() windowspecdefinition(a#10L ASC NULLS FIRST, b#11L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rowId#8], [a#10L ASC NULLS FIRST, b#11L ASC NULLS FIRST]
   +- TakeOrderedAndProject(limit=5, orderBy=[a#10L ASC NULLS FIRST,b#11L ASC NULLS FIRST], output=[a#10L,b#11L])
      +- FileScan parquet default.t1[a#10L,b#11L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/opensource/spark/spark-warehouse/org.apache.spark...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:bigint,b:bigint>
```
If remove sort. The logic is limit first and then sort:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Window [row_number() windowspecdefinition(a#10L ASC NULLS FIRST, b#11L ASC NULLS FIRST, specifiedwindowframe(RowFrame, unboundedpreceding$(), currentrow$())) AS rowId#8], [a#10L ASC NULLS FIRST, b#11L ASC NULLS FIRST]
   +- Sort [a#10L ASC NULLS FIRST, b#11L ASC NULLS FIRST], false, 0
      +- GlobalLimit 5
         +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [id=#30]
            +- LocalLimit 5
               +- FileScan parquet default.t1[a#10L,b#11L] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/opensource/spark/spark-warehouse/org.apache.spark...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<a:bigint,b:bigint>

```

Reply:
Ah, I see. I got it. Thanks. I think its better to leave some comments about why we need it here.

Reply:
Thanks for explanation @wangyum , +1 for adding some comments.

Reply:
Yes. https://github.com/apache/spark/pull/31691/commits/d1839b99e74b19a59d48d344793ca84e2b775879#diff-11264d807efa58054cca2d220aae8fba644ee0f0f2a4722c46d52828394846efR624